### ISSUE TITLE
Deadlock in Singleton.GetOrCreateAsync

### ISSUE THREAD

**Issue by tlecomte** [Azure/azure-amqp#204]
We have observed what looks like a deadlock while receiving messages.

The issue occurred in a server-side process listening to ~20 Service Bus topic subscriptions in parallel, in separate Task's. We have not exactly found what's the trigger. At some point it seemed that Internet connectivity issues might be a contributing factor, but it's not clear.

Below is a stacktrace of a Task that was stuck in this state. It's from a process dump that was taken 3 days after the deadlock started. Apparently the task did not move for these 3 days. The app is using `Azure.Messaging.ServiceBus` 7.3.0, which uses `Microsoft.Azure.Amqp` 2.5.3. The app is calling `ServiceBusMessageReceiver.ReceiveMessageAsync`, wrapped in a Polly retry policy:

```csharp
message = await RetryPolicies.SubscriptionPolicy
    .ExecuteAsync(token => messageReceiver.ReceiveMessageAsync(cancellationToken: cancellationToken), cancellationToken)
    .ConfigureAwait(false);
```

```
System.Private.CoreLib.dll!System.Threading.Tasks.Task<System.__Canon>.TrySetResult(System.__Canon result) Line 398	C#
[Async] Microsoft.Azure.Amqp.dll!Microsoft.Azure.Amqp.Singleton<Microsoft.Azure.Amqp.ReceivingAmqpLink>.GetOrCreateAsync(System.TimeSpan timeout)	Unknown
[Async] Azure.Messaging.ServiceBus.dll!Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.ReceiveMessagesAsyncInternal(int maxMessages, System.TimeSpan? maxWaitTime, System.TimeSpan timeout, System.Threading.CancellationToken cancellationToken) Line 325	C#
[Async] Azure.Messaging.ServiceBus.dll!Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.ReceiveMessagesAsync.AnonymousMethod__38_0((Azure.Messaging.ServiceBus.Amqp.AmqpReceiver, int maxMessages, System.TimeSpan? maxWaitTime) value, System.TimeSpan timeout, System.Threading.CancellationToken token) Line 290	C#
[Async] Azure.Messaging.ServiceBus.dll!Azure.Messaging.ServiceBus.ServiceBusRetryPolicy.RunOperation<(Azure.Messaging.ServiceBus.Amqp.AmqpReceiver, int, System.TimeSpan?), System.Collections.Generic.IReadOnlyList<Azure.Messaging.ServiceBus.ServiceBusReceivedMessage>>(System.Func<(Azure.Messaging.ServiceBus.Amqp.AmqpReceiver, int, System.TimeSpan?), System.TimeSpan, System.Threading.CancellationToken, System.Threading.Tasks.ValueTask<System.Collections.Generic.IReadOnlyList<Azure.Messaging.ServiceBus.ServiceBusReceivedMessage>>> operation, (Azure.Messaging.ServiceBus.Amqp.AmqpReceiver, int, System.TimeSpan?) t1, Azure.Messaging.ServiceBus.Core.TransportConnectionScope scope, System.Threading.CancellationToken cancellationToken) Line 149	C#
[Async] Azure.Messaging.ServiceBus.dll!Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.ReceiveMessagesAsync(int maxMessages, System.TimeSpan? maxWaitTime, System.Threading.CancellationToken cancellationToken) Line 294	C#
[Async] Azure.Messaging.ServiceBus.dll!Azure.Messaging.ServiceBus.ServiceBusReceiver.ReceiveMessagesAsync(int maxMessages, System.TimeSpan? maxWaitTime, bool isProcessor, System.Threading.CancellationToken cancellationToken) Line 295	C#
[Async] Azure.Messaging.ServiceBus.dll!Azure.Messaging.ServiceBus.ServiceBusReceiver.ReceiveMessagesAsync(int maxMessages, System.TimeSpan? maxWaitTime, System.Threading.CancellationToken cancellationToken) Line 249	C#
[Async] Azure.Messaging.ServiceBus.dll!Azure.Messaging.ServiceBus.ServiceBusReceiver.ReceiveMessageAsync(System.TimeSpan? maxWaitTime, System.Threading.CancellationToken cancellationToken) Line 364	C#
[Async] Polly.dll!Polly.Retry.AsyncRetryEngine.ImplementationAsync<Azure.Messaging.ServiceBus.ServiceBusReceivedMessage>(System.Func<Polly.Context, System.Threading.CancellationToken, System.Threading.Tasks.Task<Azure.Messaging.ServiceBus.ServiceBusReceivedMessage>> action, Polly.Context context, System.Threading.CancellationToken cancellationToken, Polly.ExceptionPredicates shouldRetryExceptionPredicates, Polly.ResultPredicates<Azure.Messaging.ServiceBus.ServiceBusReceivedMessage> shouldRetryResultPredicates, System.Func<Polly.DelegateResult<Azure.Messaging.ServiceBus.ServiceBusReceivedMessage>, System.TimeSpan, int, Polly.Context, System.Threading.Tasks.Task> onRetryAsync, int permittedRetryCount, System.Collections.Generic.IEnumerable<System.TimeSpan> sleepDurationsEnumerable, System.Func<int, Polly.DelegateResult<Azure.Messaging.ServiceBus.ServiceBusReceivedMessage>, Polly.Context, System.TimeSpan> sleepDurationProvider, bool continueOnCapturedContext)	Unknown
[Async] Polly.dll!Polly.AsyncPolicy.ExecuteAsync<Azure.Messaging.ServiceBus.ServiceBusReceivedMessage>(System.Func<Polly.Context, System.Threading.CancellationToken, System.Threading.Tasks.Task<Azure.Messaging.ServiceBus.ServiceBusReceivedMessage>> action, Polly.Context context, System.Threading.CancellationToken cancellationToken, bool continueOnCapturedContext)	Unknown
[Async] Entry point in our app
```

**Comment by xinchen10** [comment]
In an ideal world where everyone is using async tasks correctly, such issue would not exist. It is not specific to awaited tasks. Blocking the thread from any callbacks could potentially affect the initiator. Task just make it much easier to write such code. Well, our goal is to make scenarios work, rather than writing perfect code. I have merged the PR.
